File: app/src/main/java/com/pedrosoares/cotacaoapp/presentation/view/fragment/ExchangeListFragment.java
package com.pedrosoares.cotacaoapp.presentation.view.fragment;


import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.content.ContextCompat;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdView;
import com.pedrosoares.cotacaoapp.MainActivity;
import com.pedrosoares.cotacaoapp.R;
import com.pedrosoares.cotacaoapp.core.base.BaseFragment;
import com.pedrosoares.cotacaoapp.data.preferences.ManagerPreferences;
import com.pedrosoares.cotacaoapp.model.domain.ARSDomain;
import com.pedrosoares.cotacaoapp.model.domain.BTCDomain;
import com.pedrosoares.cotacaoapp.model.domain.CoinsDomain;
import com.pedrosoares.cotacaoapp.model.domain.EURDomain;
import com.pedrosoares.cotacaoapp.model.domain.GBPDomain;
import com.pedrosoares.cotacaoapp.model.domain.LTCDomain;
import com.pedrosoares.cotacaoapp.model.domain.USDDomain;
import com.pedrosoares.cotacaoapp.presentation.CoinsContract;
import com.pedrosoares.cotacaoapp.presentation.presenter.CoinsPresenter;
import com.pedrosoares.cotacaoapp.presentation.view.adapter.ExchangeRateAdapter;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class ExchangeListFragment extends BaseFragment<CoinsContract.CoinsListPresenter> implements CoinsContract.CoinsListView {

    //region BINDS
//    @Bind(R.id.rv_list_exchange)
//    RecyclerView rvListExchange;

//    @Bind(R.id.fragment_error)
    View includeLayoutError;

//    @Bind(R.id.iv_recycler_layout_grid)
//    ImageView imageViewOne;
//
//    @Bind(R.id.iv_recycler_layout_linear)
//    ImageView imageViewTwo;

//    @Bind(R.id.fragment_loading)
    View includeLayoutLoading;

//    @Bind(R.id.tv_last_update)
//    TextView lastUpdate;

//    @Bind(R.id.swipe_refresh)
    SwipeRefreshLayout swipeRefresh;

    ImageView imageRefreshError;

    //endregion

    private ExchangeRateAdapter exchangeRateAdapter;
    private List<Object> coinsDomainList;

    TextView lastUpdate;
    RecyclerView rvListExchange;
    AdView adView;


    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.fragment_exchange_list, container, false);

        swipeRefresh = view.findViewById(R.id.swipe_refresh);
        lastUpdate = view.findViewById(R.id.tv_last_update);
        includeLayoutError = view.findViewById(R.id.fragment_error);
        includeLayoutLoading = view.findViewById(R.id.fragment_loading);
        rvListExchange = view.findViewById(R.id.rv_list_exchange);
        imageRefreshError = view.findViewById(R.id.image_refresh);

        adView = view.findViewById(R.id.ad_view);
        AdRequest adRequest = new AdRequest.Builder().build();
        adView.loadAd(adRequest);

        return view;

    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        ButterKnife.bind(this, getActivity());

        imageRefreshError.setOnClickListener(v -> presenter.fetchCoins());

        initUi();
    }

    @Override
    public void onResume() {
        super.onResume();
        if (isConnected()) {
            if (getView() != null) {
                Snackbar.make(getView(), "Conectado", 1000).show();
                //saveUserPreferences();
            }
        } else {
            if (getView() != null)
                Snackbar.make(getView(), "Sem Conexão", 1000).show();
        }


        swipeRefresh.setOnRefreshListener(() ->  {
            swipeRefresh.setColorSchemeResources(android.R.color.holo_green_dark);
            initUi();
            swipeRefresh.setRefreshing(false);
        });

    }

    private void initUi(){

        presenter.fetchCoins();
        coinsDomainList = new ArrayList<>();
        exchangeRateAdapter = new ExchangeRateAdapter(getActivity(), coinsDomainList);
        rvListExchange.setAdapter(exchangeRateAdapter);

        LinearLayoutManager layoutManager = new LinearLayoutManager(getActivity());
        rvListExchange.setHasFixedSize(true);
        rvListExchange.setLayoutManager(layoutManager);


    }

    @Override
    public CoinsContract.CoinsListPresenter createPresenter() {
        return new CoinsPresenter(this);
    }

    @Override
    public void populateCoins(CoinsDomain coinsDomain) {
        if (getContext() != null) swipeRefresh.setBackground(ContextCompat.getDrawable(getContext(), R.drawable.background));
        addCoinsToArray(coinsDomain);
        lastUpdate.setText(getString(R.string.last_update).concat(" "+ changeDateFormat(coinsDomain.getBTC().getCreateDate())));
        exchangeRateAdapter.notifyDataSetChanged();
    }

    private void addCoinsToArray(@NonNull CoinsDomain coinsDomain) {
        ARSDomain ars = coinsDomain.getARS();
        BTCDomain btc = coinsDomain.getBTC();
        USDDomain usd = coinsDomain.getUSD();
        LTCDomain ltc = coinsDomain.getLTC();
        EURDomain eur = coinsDomain.getEUR();
        GBPDomain gbp = coinsDomain.getGBP();

        defineCardsPosition(ars, btc, usd, ltc, eur, gbp);
    }

    private String changeDateFormat(String date){
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
        SimpleDateFormat targetFormat = new SimpleDateFormat("dd-MM-yyyy HH:mm:ss");
        Date d = null;
        String dateFormatted;
        try{
            d = simpleDateFormat.parse(date);
        } catch (Exception e) {
            e.printStackTrace();
        }
        dateFormatted = targetFormat.format(d);
        return dateFormatted.replace("-","/");
    }

    private void defineCardsPosition(@NonNull ARSDomain ars,@NonNull BTCDomain btc,@NonNull USDDomain usd,@NonNull LTCDomain ltc,@NonNull EURDomain eur,@NonNull GBPDomain gbp) {
        coinsDomainList.add(0,usd);
        coinsDomainList.add(1,eur);
        coinsDomainList.add(2,gbp);
        coinsDomainList.add(3,ars);
        coinsDomainList.add(4,btc);
        coinsDomainList.add(5,ltc);
    }

    @Override
    public void success() {
        lastUpdate.setVisibility(View.VISIBLE);
        rvListExchange.setVisibility(View.VISIBLE);
        includeLayoutLoading.setVisibility(View.GONE);
        includeLayoutError.setVisibility(View.GONE);
        //includeToolbarExchange.setVisibility(View.VISIBLE);
    }

    @Override
    public void loading() {
        lastUpdate.setVisibility(View.GONE);
        includeLayoutLoading.setVisibility(View.VISIBLE);
        includeLayoutError.setVisibility(View.GONE);
        rvListExchange.setVisibility(View.GONE);
        //includeToolbarExchange.setVisibility(View.VISIBLE);
    }

    @Override
    public void error() {
        lastUpdate.setText("");
        if (getActivity() != null) swipeRefresh.setBackgroundColor(getActivity().getResources().getColor(R.color.color_white));
        includeLayoutError.setVisibility(View.VISIBLE);
        includeLayoutLoading.setVisibility(View.GONE);
        rvListExchange.setVisibility(View.GONE);
        //includeToolbarExchange.setVisibility(View.GONE);
    }


//    @OnClick(R.id.image_refresh)
//    void onClickReload(){ presenter.fetchCoins(); }


}

File: app/src/main/java/com/pedrosoares/cotacaoapp/presentation/view/fragment/ExchangeListFragment.kt
package com.pedrosoares.cotacaoapp.presentation.view.fragment


import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.content.ContextCompat
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.google.android.gms.ads.AdRequest

import com.pedrosoares.cotacaoapp.R
import com.pedrosoares.cotacaoapp.core.base.BaseFragment
import com.pedrosoares.cotacaoapp.model.domain.*
import com.pedrosoares.cotacaoapp.presentation.CurrencyContract
import com.pedrosoares.cotacaoapp.presentation.presenter.CurrencyPresenter
import com.pedrosoares.cotacaoapp.presentation.view.adapter.ExchangeRateAdapter

import java.text.SimpleDateFormat
import java.util.ArrayList
import java.util.Date

import kotlinx.android.synthetic.main.fragment_error.*
import kotlinx.android.synthetic.main.fragment_exchange_list.*

class ExchangeListFragment : BaseFragment<CurrencyContract.CurrencyListPresenter>(), CurrencyContract.CurrencyListView {

    private val exchangeRateAdapter : ExchangeRateAdapter by lazy {
        ExchangeRateAdapter(context!!, currencyDomainList)
//                , object : ExchangeRateListener {
//            override fun onCurrencySelected(bid: String) {
//                val converterFragment = ConverterFragment()
//                val bundle = Bundle()
//                bundle.putString("usd", bid)
//                converterFragment.arguments = bundle
//                converterFragment.getCurrencyDataSelected(bundle.let { converterFragment.arguments!! })
//            }
//        })

    }

    private var currencyDomainList: MutableList<Any> = ArrayList()

    override fun createPresenter() = CurrencyPresenter(this)

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_exchange_list, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val adRequest = AdRequest.Builder().build()
        adView.loadAd(adRequest)
        imageRefreshError.setOnClickListener { presenter!!.fetchCurrency() }
        initUi()
    }

    override fun onResume() {
        super.onResume()

        if (isConnected) {
            view.let{ Snackbar.make(view!!, "Conectado", 1000).show() }

        } else {
            view.let{ Snackbar.make(view!!, "Sem Conexão", 1000).show() }
        }


        swipeRefresh.setOnRefreshListener {
            swipeRefresh.setColorSchemeResources(android.R.color.holo_green_dark)
            initUi()
            swipeRefresh.isRefreshing = false
        }

    }

    private fun initUi() {

        presenter!!.fetchCurrency()

        with(rvListExchange){
            adapter = exchangeRateAdapter
            setHasFixedSize(true)
            layoutManager = LinearLayoutManager(activity)
        }
    }



    override fun populateCurrency(currencyDomain: CurrencyDomain) {
        context?.let { swipeRefresh.background = ContextCompat.getDrawable(context!!, R.drawable.background) }
        addCurrencyToArray(currencyDomain)
        tvLastUpdate.text = "${getString(R.string.last_update)} ${currencyDomain.btc?.createDate?.let {changeDateFormat(it)}}"
        exchangeRateAdapter.notifyDataSetChanged()
    }

    private fun addCurrencyToArray(currencyDomain: CurrencyDomain) {
        currencyDomain.let {
            val ars = it.ars
            val btc = it.btc
            val usd = it.usd
            val ltc = it.ltc
            val eur = it.eur
            val gbp = it.gbp

            defineCardsPosition(ars!!, btc!!, usd!!, ltc!!, eur!!, gbp!!)
        }
    }

    private fun changeDateFormat(date: String): String {
        val simpleDateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
        val targetFormat = SimpleDateFormat("dd-MM-yyyy HH:mm:ss")
        var d: Date? = null
        lateinit var dateFormatted : String
        try {
            d = simpleDateFormat.parse(date)
        } catch (e: Exception) {
            e.printStackTrace()
        }

        dateFormatted = targetFormat.format(d)
        return dateFormatted.replace("-", "/")
    }

    private fun defineCardsPosition(ars: ARSDomain, btc: BTCDomain, usd: USDDomain, ltc: LTCDomain, eur: EURDomain, gbp: GBPDomain) {
        with(currencyDomainList) {
            add(0, usd)
            add(1, eur)
            add(2, gbp)
            add(3, ars)
            add(4, btc)
            add(5, ltc)
        }
    }

    override fun success() {
        tvLastUpdate.visibility = View.VISIBLE
        rvListExchange.visibility = View.VISIBLE
        includeLayoutLoading.visibility = View.GONE
        includeLayoutError.visibility = View.GONE
    }

    override fun loading() {
        tvLastUpdate.visibility = View.GONE
        includeLayoutLoading.visibility = View.VISIBLE
        includeLayoutError.visibility = View.GONE
        rvListExchange.visibility = View.GONE
    }

    override fun error() {
        tvLastUpdate.text = ""
        if (activity != null) swipeRefresh.setBackgroundColor(activity!!.resources.getColor(R.color.color_white))
        includeLayoutError.visibility = View.VISIBLE
        includeLayoutLoading.visibility = View.GONE
        rvListExchange.visibility = View.GONE
    }




}
